File: flutter/lib/core/logic/ask_rail.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../models/jobs.dart';
import 'job_live.dart';

part 'ask_rail.freezed.dart';

/// 节点状态：`waiting` 是「排上队但还没轮到」，与还没走到的 `todo` 不是一回事。
enum RailStatus { todo, waiting, running, done, failed, cancelled }

/// 后台写库任务的状态；`unknown` 表示状态读不到，而不是没有任务。
enum KbStatus {
  queued,
  running,
  succeeded,
  failed,
  cancelled,
  unknown;

  static KbStatus of(JobStatus status) => switch (status) {
    JobStatus.queued => KbStatus.queued,
    JobStatus.running => KbStatus.running,
    JobStatus.succeeded => KbStatus.succeeded,
    JobStatus.failed => KbStatus.failed,
    JobStatus.cancelled => KbStatus.cancelled,
  };

  bool get isActive => this == KbStatus.queued || this == KbStatus.running;
}

/// 答案交付后才排队的写库任务。
@freezed
abstract class BackgroundKb with _$BackgroundKb {
  const factory BackgroundKb({
    required KbStatus status,
    @Default(0) int current,
    @Default(0) int total,
  }) = _BackgroundKb;

  /// 请求失败或状态读不到时的占位。
  static const unknown = BackgroundKb(status: KbStatus.unknown);
}

/// 阶段节点条的一格。[hint] 是节点下的一行小字，承载「（1/10 篇）」这类计数与原因。
@freezed
abstract class AskRailNode with _$AskRailNode {
  const factory AskRailNode({
    required String key,
    required String label,
    required RailStatus status,
    String? hint,
  }) = _AskRailNode;
}

AskRailNode _kbNode(BackgroundKb? kb) {
  final label = StageKey.kb.label;
  AskRailNode node(RailStatus status, String hint) => AskRailNode(
    key: StageKey.kb.name,
    label: label,
    status: status,
    hint: hint,
  );
  if (kb == null) return node(RailStatus.todo, '答案交付后在后台进行');
  final count = kb.total > 0 ? '（${kb.current}/${kb.total} 篇）' : '';
  return switch (kb.status) {
    KbStatus.queued => node(RailStatus.waiting, '等待后台，优先执行新问答$count'),
    KbStatus.running => node(RailStatus.running, '后台写入中$count'),
    KbStatus.succeeded => node(
      RailStatus.done,
      kb.total > 0 ? '已写入 ${kb.total} 篇' : '已写入',
    ),
    KbStatus.failed => node(RailStatus.failed, '写入失败，答案不受影响'),
    KbStatus.cancelled => node(RailStatus.cancelled, '已取消，答案不受影响'),
    KbStatus.unknown => node(RailStatus.todo, '状态暂时无法读取'),
  };
}

/// 答案页与运行页共用的节点序列。逐字对应 `frontend/src/components/ask/askRail.ts`。
///
/// 写库不在问答流水线里跑：API 侧 `run_ask(..., defer_kb=True)`，答案交付之后
/// worker 才另建 `answer_kb` 后台任务，所以它是「综合成稿」之后的节点，
/// 状态只能来自那个子任务，问答的 SSE 里永远不会有 kb 阶段事件。
///
/// [settled] 表示答案已出：重新进入答案页时 SSE 状态是空的，
/// 前序阶段一律按已完成渲染，否则节点会全灰。
List<AskRailNode> askRailNodes(
  JobLive live, {
  required bool useKb,
  BackgroundKb? kb,
  bool settled = false,
}) {
  final nodes = <AskRailNode>[
    for (final stage in StageKey.askPipeline) _stageNode(live, stage, settled),
  ];
  return useKb ? [...nodes, _kbNode(kb)] : nodes;
}

AskRailNode _stageNode(JobLive live, StageKey stage, bool settled) {
  final state = live.stages[stage];
  final finished = state?.status == StageStatus.finished;
  // 已完成的阶段把详情摘要（「相关 4/6」之类）挂在节点小字上。
  final summary = finished ? stageSummary(stage, state!.detail) : '';
  return AskRailNode(
    key: stage.name,
    label: stage.label,
    status: finished || settled
        ? RailStatus.done
        : (state != null ? RailStatus.running : RailStatus.todo),
    hint: summary.isEmpty ? null : summary,
  );
}

File: apple/PicoSeekKit/Sources/PicoSeekKit/Logic/AskRail.swift
import Foundation

/// 节点状态：`waiting` 是「排上队但还没轮到」，与还没走到的 `todo` 不是一回事。
public enum RailStatus: Sendable, Hashable {
    case todo
    case waiting
    case running
    case done
    case failed
    case cancelled
}

/// 流水线上的一个节点。视图只负责把它画出来，状态判定全在这里。
public struct RailNode: Sendable, Hashable, Identifiable {
    public var key: String
    public var label: String
    public var status: RailStatus
    /// 节点下的一行小字，用来承载「（1/10 篇）」这类计数与原因。
    public var hint: String?

    public var id: String { key }

    public init(key: String, label: String, status: RailStatus, hint: String? = nil) {
        self.key = key
        self.label = label
        self.status = status
        self.hint = hint
    }
}

/// 答案交付后才排队的写库任务；`unknown` 表示状态读不到，而不是没有任务。
public struct BackgroundKb: Sendable, Hashable {
    public enum Status: Sendable, Hashable {
        case queued
        case running
        case succeeded
        case failed
        case cancelled
        case unknown
    }

    public var status: Status
    public var current: Int
    public var total: Int

    public init(status: Status, current: Int = 0, total: Int = 0) {
        self.status = status
        self.current = current
        self.total = total
    }

    /// 后台子任务（`kind == "answer_kb"`）的进度：`progress = {stage:"kb", current, total}`。
    public init(job: Job) {
        self.init(
            status: Status(job.status),
            current: job.progress?.current ?? 0,
            total: job.progress?.total ?? 0
        )
    }
}

public extension BackgroundKb.Status {
    init(_ status: JobStatus) {
        switch status {
        case .queued: self = .queued
        case .running: self = .running
        case .succeeded: self = .succeeded
        case .failed: self = .failed
        case .cancelled: self = .cancelled
        }
    }
}

/// 写库不在问答流水线里跑：API 侧 `defer_kb=True`，答案交付后 worker 才建一个
/// `answer_kb` 后台任务，所以它是「综合成稿」之后的附加节点，状态来自那个独立任务。
private let kbLabel = "写入知识库"

private func kbNode(_ kb: BackgroundKb?) -> RailNode {
    guard let kb else {
        return RailNode(key: "kb", label: kbLabel, status: .todo, hint: "答案交付后在后台进行")
    }
    let count = kb.total > 0 ? "（\(kb.current)/\(kb.total) 篇）" : ""
    switch kb.status {
    case .queued:
        return RailNode(key: "kb", label: kbLabel, status: .waiting, hint: "等待后台，优先执行新问答\(count)")
    case .running:
        return RailNode(key: "kb", label: kbLabel, status: .running, hint: "后台写入中\(count)")
    case .succeeded:
        return RailNode(
            key: "kb",
            label: kbLabel,
            status: .done,
            hint: kb.total > 0 ? "已写入 \(kb.total) 篇" : "已写入"
        )
    case .failed:
        return RailNode(key: "kb", label: kbLabel, status: .failed, hint: "写入失败，答案不受影响")
    case .cancelled:
        return RailNode(key: "kb", label: kbLabel, status: .cancelled, hint: "已取消，答案不受影响")
    case .unknown:
        return RailNode(key: "kb", label: kbLabel, status: .todo, hint: "状态暂时无法读取")
    }
}

/// 答案页与运行页共用的节点序列；`settled` 表示答案已出，前序阶段一律算完成
/// （重新进入答案页时 SSE 实时状态是空的，不能让节点全灰）。
/// 逐字移植 `frontend/src/components/ask/askRail.ts`。
public func askRailNodes(
    live: JobLive,
    useKb: Bool,
    kb: BackgroundKb? = nil,
    settled: Bool = false
) -> [RailNode] {
    let nodes = StageKey.askPipeline.map { stage in
        RailNode(
            key: stage.rawValue,
            label: stage.label,
            status: {
                if settled || live.stages[stage]?.status == .finished { return .done }
                return live.stages[stage] == nil ? .todo : .running
            }()
        )
    }
    return useKb ? nodes + [kbNode(kb)] : nodes
}
